File: src/jastudio_src/jastudio/note/vocabulary/vocabnote_register.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]

from jastudio.note.notefields.tag_flag_field import TagFlagField
from jastudio.note.tags import Tags

if TYPE_CHECKING:
    from jastudio.note.vocabulary.vocabnote import VocabNote
    from jastudio.sysutils.weak_ref import WeakRef

# todo performance: memory: high-priority: combine into a single bitfield in memory
class VocabNoteRegister(Slots):
    def __init__(self, vocab: WeakRef[VocabNote]) -> None:
        self._vocab: WeakRef[VocabNote] = vocab

    @property
    def polite(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.polite)
    @property
    def formal(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.formal)
    @property
    def informal(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.informal)
    @property
    def archaic(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.archaic)
    @property
    def sensitive(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.sensitive)
    @property
    def vulgar(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.vulgar)
    @property
    def humble(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.humble)
    @property
    def literary(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.literary)
    @property
    def honorific(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.honorific)
    @property
    def rough_masculine(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.rough_masculine)
    @property
    def soft_feminine(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.soft_feminine)
    @property
    def slang(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.slang)
    @property
    def derogatory(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.derogatory)
    @property
    def childish(self) -> TagFlagField: return TagFlagField(self._vocab, Tags.Vocab.Register.childish)

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/VocabNoteRegister.cs
using JAStudio.Core.Note.NoteFields;

namespace JAStudio.Core.Note.Vocabulary;

public class VocabNoteRegister
{
   public VocabNoteRegister(VocabNote vocab) => Vocab = vocab;

   VocabNote Vocab { get; }

   public TagFlagField Polite => new(Vocab, Tags.Vocab.Register.Polite);
   public TagFlagField Formal => new(Vocab, Tags.Vocab.Register.Formal);
   public TagFlagField Informal => new(Vocab, Tags.Vocab.Register.Informal);
   public TagFlagField Archaic => new(Vocab, Tags.Vocab.Register.Archaic);
   public TagFlagField Sensitive => new(Vocab, Tags.Vocab.Register.Sensitive);
   public TagFlagField Vulgar => new(Vocab, Tags.Vocab.Register.Vulgar);
   public TagFlagField Humble => new(Vocab, Tags.Vocab.Register.Humble);
   public TagFlagField Literary => new(Vocab, Tags.Vocab.Register.Literary);
   public TagFlagField Honorific => new(Vocab, Tags.Vocab.Register.Honorific);
   public TagFlagField RoughMasculine => new(Vocab, Tags.Vocab.Register.RoughMasculine);
   public TagFlagField SoftFeminine => new(Vocab, Tags.Vocab.Register.SoftFeminine);
   public TagFlagField Slang => new(Vocab, Tags.Vocab.Register.Slang);
   public TagFlagField Derogatory => new(Vocab, Tags.Vocab.Register.Derogatory);
   public TagFlagField Childish => new(Vocab, Tags.Vocab.Register.Childish);
}
